lib/bb/data_smart.py

```diff
@@ -1,351 +1,1045 @@
+        the files in which they were added.
+        """
+        d = self.dataroot
+        history = self.variable(var)
+        finalitems = (d.getVar(var) or '').split()
+        filemap = {}
+        isset = False
+        for event in history:
+            if 'flag' in event:
+                continue
+            if event['op'] == ':remove':
+                continue
+            if isset and event['op'] == 'set?':
+                continue
+            isset = True
+            items = d.expand(event['detail']).split()
+            for item in items:
+                # This is a little crude but is belt-and-braces to avoid us
+                # having to handle every possible operation type specifically
+                if item in finalitems and not item in filemap:
+                    filemap[item] = event['file']
+        return filemap
+
+    def del_var_history(self, var, f=None, line=None):
+        """If file f and line are not given, the entire history of var is deleted"""
+        if var in self.variables:
+            if f and line:
+                self.variables[var] = [ x for x in self.variables[var] if x['file']!=f and x['line']!=line]
+            else:
+                self.variables[var] = []
 
-class DataSmart:
+class DataSmart(MutableMapping):
     def __init__(self):
         self.dict = {}
 
-    def expand(self,s, varname):
-        def var_sub(match):
-            key = match.group()[2:-1]
-            if varname and key:
-                if varname == key:
-                    raise Exception("variable %s references itself!" % varname)
-            var = self.getVar(key, 1)
-            if var is not None:
-                return var
-            else:
-                return match.group()
+        self.inchistory = IncludeHistory()
+        self.varhistory = VariableHistory(self)
+        self._tracking = False
+
+        self.expand_cache = {}
+
+        # cookie monster tribute
+        # Need to be careful about writes to overridedata as
+        # its only a shallow copy, could influence other data store
+        # copies!
+        self.overridedata = {}
+        self.overrides = None
+        self.overridevars = set(["OVERRIDES", "FILE"])
+        self.inoverride = False
+
+    def enableTracking(self):
+        self._tracking = True
 
-        def python_sub(match):
-            import bb
-            code = match.group()[3:-1]
-            locals()['d'] = self
-            s = eval(code)
-            if type(s) == types.IntType: s = str(s)
-            return s
+    def disableTracking(self):
+        self._tracking = False
 
-        if type(s) is not types.StringType: # sanity check
-            return s
+    def expandWithRefs(self, s, varname):
 
-        while s.find('$') != -1:
+        if not isinstance(s, str): # sanity check
+            return VariableParse(varname, self, s)
+
+        varparse = VariableParse(varname, self)
+
+        while s.find('${') != -1:
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

It appears that you do not need the position returned by the [`find()`](https://docs.python.org/3/library/stdtypes.html#str.find) function. Consider using the operators [`in` and `not in`](https://docs.python.org/3/reference/expressions.html#membership-test-details) to check for simple membership, or more specific operations [`startswith()`](https://docs.python.org/3/library/stdtypes.html#str.startswith) and [`endswith()`](https://docs.python.org/3/library/stdtypes.html#str.endswith) to check for prefix and suffix relationships with the string.